File: SharedPreferences/java/MainActivity.java
package com.example.sharedpreferences;

import androidx.appcompat.app.AppCompatActivity;

import android.content.SharedPreferences;
import android.os.Bundle;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

public class MainActivity extends AppCompatActivity {

    private EditText mEditText;
    private Button mButton;
    private SharedPreferences pref;
    private SharedPreferences.Editor editor;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        //获取SharedPreferences实例
        pref = getSharedPreferences("test",MODE_PRIVATE);

        mButton = (Button)findViewById(R.id.save);
        mEditText = (EditText)findViewById(R.id.editText);

        mButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                //edit()获取SharedPreferences.Editor实例，再使用putString()添加数据，最后使用apply()提交
                editor = pref.edit();
                editor.putString("editText", mEditText.getText().toString());
                editor.apply();
            }
        });

        //当活动创建的时候检查时候是否有已经存储的数据，如果有就将数据呈现在EditText实例中
        String str = pref.getString("editText","");
        if(!TextUtils.isEmpty(str)){        //使用TextUtils.isEmpty()判断字符串是否为null或者空字符串
            mEditText.setText(str);
        }
    }
}

File: WidgetDemo1/java/MainActivity.kt
package org.anvei.widgettest

import android.content.Context
import android.os.Bundle
import android.view.*
import android.widget.Button
import android.widget.PopupWindow
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        popup.setOnClickListener {
            val contentView: View = LayoutInflater.from(this).inflate(R.layout.pop_window_item, null)
            val popupWindow = PopupWindow(contentView, ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT, true)
            with(popupWindow) {
                showAtLocation(it, Gravity.BOTTOM, 0, 0)
            }
            contentView.findViewById<Button>(R.id.popup_cancel).setOnClickListener {
                popupWindow.dismiss()
            }
            contentView.findViewById<Button>(R.id.popup_select).setOnClickListener {
                Toast.makeText(this, "select photo", Toast.LENGTH_SHORT).show()
            }
            contentView.findViewById<Button>(R.id.popup_take_photo).setOnClickListener {
                Toast.makeText(this, "take photo", Toast.LENGTH_SHORT).show()
            }
        }

//        val floatingButton = Button(this)
//        floatingButton.text = "FloatingButton"
//        val layoutParams = WindowManager.LayoutParams()
//
//        layoutParams.apply {
//            height = WindowManager.LayoutParams.WRAP_CONTENT
//            width = WindowManager.LayoutParams.WRAP_CONTENT
//            flags = WindowManager.LayoutParams.FLAG_NOT_TOUCH_MODAL or
//                    WindowManager.LayoutParams.FLAG_NOT_FOCUSABLE
//            gravity = Gravity.START or Gravity.TOP
//            x = 100
//            y = 300
//        }
//        (getSystemService(Context.WINDOW_SERVICE) as WindowManager).addView(floatingButton, layoutParams)

    }
}